code/modules/urist/modules/shipbattles/shipweapons/shipweapons2.dm

```diff
@@ -0,0 +1,383 @@
+#define WEAPON_STATUS_RECHARGING 1 //weapon recharging
+#define WEAPON_STATUS_RELOADING 2 //weapon is chambering a round from internal magazine
+#define WEAPON_STATUS_REQUIRES_RELOAD 3 //weapon requires manual reload.
+#define WEAPON_STATUS_OFFLINE 4 //weapon is offline
+#define WEAPON_STATUS_READY 5 //weapon is ready to fire
+#define WEAPON_STATUS_FIRING 6 //weapon is firing
+
+#define WEAPON_FIREMODE_SINGLE 7 //fires one round.
+#define WEAPON_FIREMODE_SALVO 8 //fires rounds one after the other, cycles from magazine.
+#define WEAPON_FIREMODE_BURST 9 //fires *all* chambered rounds at once.
+
+/obj/machinery/shipweapons/shipweapons2
+	icon = 'icons/urist/structures&machinery/railgun.dmi'
+	name = "shipweapontwo"
+	desc = "You shouldn't see this. Someone messed up."
+
+	var/obj/item/shipweapons/ammo/chambered //chambered munition, if only firing one at a time.
+
+	var/chamber_sound = 'sound/effects/extout.ogg'
+	var/firing_state //animation for firing
+
+	var/accepted_ammo_type = /obj/item/shipweapons/ammo //accepted ammo type.
+
+	var/cooldown = 1 SECOND //cooldown, functionally used in Fire()
+	var/magazine_capacity = 0 //internal mag capacity
+	var/chamber_delay = 5 SECONDS //delay between firing and loading another round.
+	var/chamber_capacity = 1 //if firing multiple munitions, how many?
+	var/salvo_rounds = 0 //used in WEAPON_FIREMODE_SALVO to determine how many shots are fired.
+	var/salvo_rounds_left = 0 //used in WEAPON_FIREMODE_SALVO
+
+	var/fires_multiple_rounds = FALSE //do we fire more than one round
+	var/has_magazine = FALSE //do we have a magazine?
+	var/supports_custom_poweruse = FALSE //do we support our rounds using a custom amount of power and recharge time?
+	var/automatic_reload = FALSE //requires manual cycling
+	var/supports_multiple_fire_modes = FALSE //do we support on-the-fly switching of firemodes?
+	var/has_ammo_states = FALSE //icon state changes on ammo status
+
+	var/firemode = WEAPON_FIREMODE_SINGLE //weapon fire mode
+
+	var/list/magazine = list() //internal magazine.
+	var/list/chambered_munitions = list() //in case we have weapons that fire more than one round at once.
+
+
+
+/obj/machinery/shipweapons/shipweapons2/Charging() //completely redoing this as well to integrate things.
+	if(stat & (BROKEN|NOPOWER)) //we're broken, no charging for us.
+		return
+
+	if(supports_custom_poweruse)
+		if(chambered && chambered.custom_power_draw == TRUE)
```

Comment:
You can change this to `if(chambered?.custom_power_draw)`

It does the same thing and has better performance. any instance where you do `if(x && x.y)` you can use the `?.` operator instead.